DESIGN.md

```diff
@@ -0,0 +1,126 @@
+
+- **Read only:** this role can get job output, but not create or delete jobs. This is useful for automated tooling such as status indicators and job notifiers.
+- **Read/Write access:** this role can get job output as well as create and delete jobs. This role is useful for administrators and any kind of deployment automation that would leverage this service.
+
+![visual representation of the multiple client CA model](doc/client_cas.png)
+
+## API
+### Endpoints
+The API exposed by the RPC service will contain the following endpoints. Any errors mentioned refer to the [gRPC status codes](https://developers.google.com/maps-booking/reference/grpc-api/status_codes).
+
+#### **CreateJob**
+Receives a Command and returns a Job. Can return the following errors:
+
+- PERMISSION_DENIED: the client certificate is not signed by a CA that permits the read/write role
+- ABORTED: the command failed to launch
+
+#### **DeleteJob**
+Receives a JID and returns a DeleteJobOutput. Can return the following errors:
+
+- PERMISSION_DENIED: the client certificate is not signed by a CA that permits the read/write role
+- NOT_FOUND: the JID was not found in the process cache
+
+After DeleteJob is run, the Job will be completely deallocated from the RPC system.
+
+#### **ListJobs**
+Returns a JobList. 
+
+#### **GetJob**
+Receives a JID and returns a Job (including updated job output and execution time).
+
+### Datatypes
+- **JID**
+    Single integer representing the ID of the Job as tracked by the RPC subsystem. Custom Job IDs were added to the implementation because it could not be assured that every Job would have a unique PID for the underlying process. 
+- **Command**
+    String that holds a shell command to be executed on the host system
+- **Status**
+    Enumeration that can hold the following values:
+    - Created
+    - In Process
+    - Exited
+    - Error
+- **Job**
+    - Command (string):
+    Command holds a valid shell command that is run on the underlying server shell.   
+    - JID (integer):
+    JID holds an integer representing job ID. 
+    - CreatedTime (string):
+    Timestamp that represents when the process was launched on the server
+    - JobExecutionTime (string):
+    Timestamp that represents how long the underlying process has been running
+    - FinishedTime (string):
+    Timestamp that represents when a Job finished
+    - Status (status):
+    Instance of status enumeration. Represents state of job
+    - Output (string):
+    Output contains the output of the job at time it is retrieved
+- **JobList**
+    - Jobs ([]Job):
+    A list of Job obejcts as defined above in this list
+- **DeleteJobOutput**
+    - Job (Job):
+    Job object as defined above in this list
+    - Exit (integer):
+    Exit code as returned from the underlying process
+    - State (status):
```

Comment:
what's the difference between this and the status included in the job object?

Reply:
This one was superfluous. We can just use the one in the job object.